kore/src/Kore/JsonRpc.hs

```diff
@@ -0,0 +1,270 @@
+    { state :: !Text
+    , patterns :: ![Text]
+    , reason :: !ReasonForHalting
+    }
+    deriving stock (Generic, Show, Eq)
+    deriving
+        (ToJSON)
+        via CustomJSON '[OmitNothingFields, FieldLabelModifier '[CamelToKebab]] ExecuteResult
+
+newtype StepResult = StepResult
+    { states :: [StepState]
+    }
+    deriving stock (Generic, Show, Eq)
+    deriving
+        (ToJSON)
+        via CustomJSON '[OmitNothingFields, FieldLabelModifier '[CamelToKebab]] StepResult
+
+data ImpliesResult = ImpliesResult
+    { satisfiable :: !Bool
+    , substitution :: !(Maybe Text)
+    }
+    deriving stock (Generic, Show, Eq)
+    deriving
+        (ToJSON)
+        via CustomJSON '[OmitNothingFields, FieldLabelModifier '[CamelToKebab]] ImpliesResult
+
+newtype SimplifyResult = SimplifyResult
+    { state :: Text
+    }
+    deriving stock (Generic, Show, Eq)
+    deriving
+        (ToJSON)
+        via CustomJSON '[OmitNothingFields, FieldLabelModifier '[CamelToKebab]] SimplifyResult
+
+data ReqOrRes = Req | Res
+
+data APIMethods = ExecuteM | StepM | ImpliesM | SimplifyM
+
+type family APIPayload (api :: APIMethods) (r :: ReqOrRes) where
+    APIPayload 'ExecuteM 'Req = ExecuteRequest
+    APIPayload 'ExecuteM 'Res = ExecuteResult
+    APIPayload 'StepM 'Req = StepRequest
+    APIPayload 'StepM 'Res = StepResult
+    APIPayload 'ImpliesM 'Req = ImpliesRequest
+    APIPayload 'ImpliesM 'Res = ImpliesResult
+    APIPayload 'SimplifyM 'Req = SimplifyRequest
+    APIPayload 'SimplifyM 'Res = SimplifyResult
+
+data API (r :: ReqOrRes) where
+    Execute :: APIPayload 'ExecuteM r -> API r
+    Step :: APIPayload 'StepM r -> API r
+    Implies :: APIPayload 'ImpliesM r -> API r
+    Simplify :: APIPayload 'SimplifyM r -> API r
+    Cancel :: API 'Req
+
+deriving stock instance Show (API 'Req)
+deriving stock instance Show (API 'Res)
+deriving stock instance Eq (API 'Req)
+deriving stock instance Eq (API 'Res)
+
+instance ToJSON (API 'Res) where
+    toJSON = \case
+        Execute payload -> toJSON payload
+        Step payload -> toJSON payload
+        Implies payload -> toJSON payload
+        Simplify payload -> toJSON payload
+
+respond :: MonadIO m => Respond (API 'Req) m (API 'Res)
+respond = \case
+    Execute _ -> pure $ Right $ Execute undefined
+    Step StepRequest{} -> pure $ Right $ Step $ StepResult []
+    Implies _ -> pure $ Right $ Implies undefined
+    Simplify _ -> pure $ Right $ Simplify undefined
+    -- this case is only reachable if the cancel appeared as part of a batch request
+    Cancel -> pure $ Left $ ErrorObj "Cancel request unsupported in batch mode" (-32001) Null
+
+runServer :: IO ()
+runServer = do
+    runStderrLoggingT $ do
+        let ss = serverSettings 31337 "*"
+        jsonrpcTCPServer V2 False ss srv
+
+srv :: MonadLoggerIO m => JSONRPCT m ()
+srv = do
+    reqQueue <- liftIO $ atomically newTChan
+    qs <- ask
+    let sendResponses r = runStderrLoggingT $ runReaderT (sendBatchResponse r) qs
```

Comment:
Can some of the `let` binding be moved to the `where` clause? I think it's nicer to see the main logic in the body of the top level function and abstract away implementation details to the `where` clause.